File: lib/models/study_plan.dart
import 'enums.dart';

/// A single day's assignment in a study plan.
class StudyDay {
  const StudyDay({
    required this.dayNumber,
    required this.date,
    required this.questionIds,
    required this.isReviewDay,
  });

  final int dayNumber;
  final DateTime date;

  /// Question ids (within the plan's [StudyPlan.version]) to learn/review today.
  final List<int> questionIds;

  /// Review days re-cover previously assigned questions rather than new ones.
  final bool isReviewDay;

  Map<String, dynamic> toJson() => {
    'dayNumber': dayNumber,
    'date': date.toIso8601String(),
    'questionIds': questionIds,
    'isReviewDay': isReviewDay,
  };

  factory StudyDay.fromJson(Map<String, dynamic> json) => StudyDay(
    dayNumber: json['dayNumber'] as int,
    date: DateTime.parse(json['date'] as String),
    questionIds: (json['questionIds'] as List).cast<int>(),
    isReviewDay: json['isReviewDay'] as bool,
  );
}

/// A dated plan that spreads all questions across the days until the test,
/// with periodic review days.
class StudyPlan {
  const StudyPlan({
    required this.version,
    required this.createdAt,
    required this.testDate,
    required this.days,
    required this.completedDayNumbers,
  });

  final TestVersion version;
  final DateTime createdAt;
  final DateTime testDate;
  final List<StudyDay> days;
  final Set<int> completedDayNumbers;

  int get totalDays => days.length;
  int get completedCount => completedDayNumbers.length;
  double get progress => totalDays == 0 ? 0 : completedCount / totalDays;

  StudyPlan copyWith({Set<int>? completedDayNumbers}) => StudyPlan(
    version: version,
    createdAt: createdAt,
    testDate: testDate,
    days: days,
    completedDayNumbers: completedDayNumbers ?? this.completedDayNumbers,
  );

  Map<String, dynamic> toJson() => {
    'version': version.storageKey,
    'createdAt': createdAt.toIso8601String(),
    'testDate': testDate.toIso8601String(),
    'days': days.map((d) => d.toJson()).toList(),
    'completedDayNumbers': completedDayNumbers.toList(),
  };

  factory StudyPlan.fromJson(Map<String, dynamic> json) => StudyPlan(
    version: json['version'] == 'v2020' ? TestVersion.v2020 : TestVersion.v2008,
    createdAt: DateTime.parse(json['createdAt'] as String),
    testDate: DateTime.parse(json['testDate'] as String),
    days: (json['days'] as List)
        .map((d) => StudyDay.fromJson(d as Map<String, dynamic>))
        .toList(),
    completedDayNumbers: (json['completedDayNumbers'] as List)
        .cast<int>()
        .toSet(),
  );
}

File: backend/src/CivicsPrep.Domain/Entities/StudyPlan.cs
using CivicsPrep.Domain.Enums;

namespace CivicsPrep.Domain.Entities;

/// <summary>
/// A dated plan that spreads all of a test's questions across the days until the exam, with
/// periodic review days.
/// </summary>
public class StudyPlan
{
    public Guid Id { get; set; } = Guid.NewGuid();

    public string UserId { get; set; } = string.Empty;

    public TestVersion Version { get; set; }

    public DateOnly TestDate { get; set; }
    public bool SeniorOnly { get; set; }

    public DateTimeOffset CreatedAt { get; set; } = DateTimeOffset.UtcNow;

    public List<StudyPlanDay> Days { get; set; } = [];

    public int TotalDays => Days.Count;
    public int CompletedCount => Days.Count(d => d.IsComplete);
    public double Progress => TotalDays == 0 ? 0 : (double)CompletedCount / TotalDays;
}

/// <summary>A single day's assignment in a <see cref="StudyPlan"/>.</summary>
public class StudyPlanDay
{
    public int Id { get; set; }

    public Guid StudyPlanId { get; set; }
    public StudyPlan? StudyPlan { get; set; }

    public int DayNumber { get; set; }
    public DateOnly Date { get; set; }

    /// <summary>Review days revisit previously assigned questions rather than new ones.</summary>
    public bool IsReviewDay { get; set; }

    public bool IsComplete { get; set; }

    public List<StudyPlanDayQuestion> Questions { get; set; } = [];
}

public class StudyPlanDayQuestion
{
    public int Id { get; set; }

    public int StudyPlanDayId { get; set; }
    public StudyPlanDay? StudyPlanDay { get; set; }

    public int QuestionNumber { get; set; }
    public int Ordinal { get; set; }
}
